File: CloudServer/stress_test.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
任务 10：高并发压测脚本
每个连接：建连 -> 发 FLUSH_DIR(type=23) 请求 -> 收响应(type=24) -> 断开
请求无需登录鉴权，服务端必回包，适合大批量压测，不污染数据库。

用法:
  python3 stress_test.py [连接数] [每连接请求数] [host] [port]
示例:
  python3 stress_test.py 100            # 100 并发连接，每连接 1 个请求
  python3 stress_test.py 300 20         # 300 连接，每连接连发 20 个请求
  python3 stress_test.py 1000 1 127.0.0.1 8888
"""
import socket
import struct
import threading
import time
import sys

PDU_HEADER = 76          # sizeof(PDU)
TYPE_FLUSH_DIR_REQ = 23  # ENUM_MSG_TYPE_FLUSH_DIR_RESPEST
TYPE_FLUSH_DIR_RSP = 24  # ENUM_MSG_TYPE_FLUSH_DIR_RESPONSE


def recv_exact(s, n):
    buf = b''
    while len(buf) < n:
        chunk = s.recv(n - len(buf))
        if not chunk:
            break
        buf += chunk
    return buf


def build_flush_pdu(path=b'./test1\x00'):
    msglen = len(path)
    total = PDU_HEADER + msglen
    # 布局: [uiPDULen_][uiMsgLen_][uiMsgType_][caData 64][caMsg]
    return struct.pack('<III', total, msglen, TYPE_FLUSH_DIR_REQ) + b'\x00' * 64 + path


def worker(idx, host, port, repeat, timeout, results):
    ok = 0
    fail = 0
    # 阶段1：连接（区分 connect 超时 —— 服务端 backlog/accept 问题）
    try:
        s = socket.create_connection((host, port), timeout=timeout)
    except Exception as e:
        results[idx] = ('conn-err', 1, 0, str(e)[:80])
        return
    s.settimeout(timeout)
    # 阶段2：请求/响应（区分 recv 超时 —— 连接被接受但请求没被处理）
    try:
        for _ in range(repeat):
            s.sendall(build_flush_pdu())
            hdr = recv_exact(s, 4)
            if len(hdr) != 4:
                fail += 1
                break
            (total,) = struct.unpack('<I', hdr)
            body = recv_exact(s, total - 4)
            if len(body) != total - 4:
                fail += 1
                break
            msgtype = struct.unpack('<I', body[4:8])[0]  # 响应包偏移: body[0:4]=msglen, body[4:8]=type
            if msgtype == TYPE_FLUSH_DIR_RSP:
                ok += 1
            else:
                fail += 1
        s.close()
    except Exception as e:
        results[idx] = ('recv-err', 1, ok, str(e)[:80])
        return
    results[idx] = ('ok', fail, ok, '')


def main():
    nconn = int(sys.argv[1]) if len(sys.argv) > 1 else 100
    repeat = int(sys.argv[2]) if len(sys.argv) > 2 else 1
    host = sys.argv[3] if len(sys.argv) > 3 else '127.0.0.1'
    port = int(sys.argv[4]) if len(sys.argv) > 4 else 8888
    timeout = float(sys.argv[5]) if len(sys.argv) > 5 else 10.0

    print(f"[开始] {nconn} 连接 x {repeat} 请求 -> {host}:{port} (超时 {timeout}s)")
    results = [None] * nconn
    threads = []
    t0 = time.time()
    for i in range(nconn):
        t = threading.Thread(target=worker, args=(i, host, port, repeat, timeout, results))
        threads.append(t)
    for t in threads:
        t.start()
    for t in threads:
        t.join()
    dur = time.time() - t0

    conn_ok = sum(1 for r in results if r and r[0] == 'ok')
    errs = [r for r in results if r and r[0] != 'ok']
    ok_req = sum(r[2] for r in results if r)
    fail_req = nconn * repeat - ok_req

    print(f"[结果] 连接成功={conn_ok}/{nconn}")
    print(f"[结果] 请求成功={ok_req}/{nconn * repeat}  请求失败={fail_req}")
    print(f"[结果] 总耗时={dur:.2f}s  连接速率≈{nconn / dur:.0f}/s  请求速率≈{ok_req / dur:.0f}/s")
    if errs:
        print("[错误样例]")
        for r in errs[:5]:
            print("   ", r)


if __name__ == '__main__':
    main()

File: CloudServer/stress_test.cpp
// stress_test.cpp — 高并发压测程序（与 stress_test.py 功能一致）
// 每个连接：建连 -> 发 FLUSH_DIR(type=23) 请求 -> 收响应(type=24) -> 统计速率
// 运行：./stress_test [连接数] [每连接请求数] [host] [port] [超时秒]
#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <unistd.h>
#include <chrono>
#include <cstdint>
#include <cstring>
#include <iostream>
#include <string>
#include <thread>
#include <vector>

static const int PDU_HEADER = 76;              // PDU 头长度，与 protocol.h 一致
static const uint32_t TYPE_FLUSH_DIR_REQ = 23; // 请求类型
static const uint32_t TYPE_FLUSH_DIR_RSP = 24; // 响应类型

struct Result { bool connErr = false; int ok = 0; };

// 构造 FLUSH_DIR 请求包：[总长][消息长][类型][caData 64][caMsg]
static std::string buildFlushPdu() {
    std::string path("./test1\0", 8);          // './test1' + '\0'，与 py 版一致
    uint32_t msglen = (uint32_t)path.size();
    uint32_t total  = PDU_HEADER + msglen;
    char hdr[12];
    memcpy(hdr,     &total,             4);    // 小端写总长（同 struct.pack('<I')）
    memcpy(hdr + 4, &msglen,            4);    // 小端写消息长
    memcpy(hdr + 8, &TYPE_FLUSH_DIR_REQ, 4);   // 小端写类型
    return std::string(hdr, 12) + std::string(64, '\0') + path;
}

// 精确接收 n 字节（循环收，处理粘包/半包）
static size_t recvExact(int fd, char* buf, size_t n) {
    size_t got = 0;
    while (got < n) {
        ssize_t r = recv(fd, buf + got, n - got, 0);
        if (r <= 0) break;
        got += (size_t)r;
    }
    return got;
}

// 单个连接的工作：建连 -> repeat 次请求/响应
static void worker(const std::string& host, int port, int repeat, int timeoutSec, Result& r) {
    int fd = socket(AF_INET, SOCK_STREAM, 0);
    if (fd < 0) { r.connErr = true; return; }
    struct timeval tv = { timeoutSec, 0 };
    setsockopt(fd, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv));  // 收包超时
    setsockopt(fd, SOL_SOCKET, SO_SNDTIMEO, &tv, sizeof(tv));  // 发包超时
    sockaddr_in addr{};
    addr.sin_family = AF_INET;
    addr.sin_port   = htons((uint16_t)port);
    inet_pton(AF_INET, host.c_str(), &addr.sin_addr);
    if (connect(fd, (sockaddr*)&addr, sizeof(addr)) != 0) {    // 阶段1：建连
        close(fd); r.connErr = true; return;
    }
    std::string pdu = buildFlushPdu();
    char buf[256];
    for (int i = 0; i < repeat; ++i) {                          // 阶段2：请求/响应
        if (send(fd, pdu.data(), (int)pdu.size(), 0) <= 0) break;
        if (recvExact(fd, buf, 4) != 4) break;                  // 先读 4 字节总长
        uint32_t total;
        memcpy(&total, buf, 4);
        if (recvExact(fd, buf, total - 4) != total - 4) break;  // 再读剩余包体
        uint32_t msgtype;
        memcpy(&msgtype, buf + 4, 4);                           // body[4:8] = 类型
        if (msgtype == TYPE_FLUSH_DIR_RSP) ++r.ok;              // 收到正确响应
    }
    close(fd);
}

int main(int argc, char** argv) {
    int nconn      = argc > 1 ? atoi(argv[1]) : 100;
    int repeat     = argc > 2 ? atoi(argv[2]) : 1;
    std::string host = argc > 3 ? argv[3] : "127.0.0.1";
    int port       = argc > 4 ? atoi(argv[4]) : 8888;
    int timeoutSec = argc > 5 ? atoi(argv[5]) : 10;

    std::cout << "[开始] " << nconn << " 连接 x " << repeat << " 请求 -> "
              << host << ":" << port << " (超时 " << timeoutSec << "s)\n";
    std::vector<Result> results(nconn);
    std::vector<std::thread> threads;
    auto t0 = std::chrono::steady_clock::now();
    for (int i = 0; i < nconn; ++i)                             // 并发开 N 个连接
        threads.emplace_back(worker, host, port, repeat, timeoutSec, std::ref(results[i]));
    for (auto& t : threads) t.join();
    double dur = std::chrono::duration<double>(std::chrono::steady_clock::now() - t0).count();

    int connOk = 0, okReq = 0;
    for (auto& r : results) { if (!r.connErr) ++connOk; okReq += r.ok; }
    int failReq = nconn * repeat - okReq;
    std::cout << "[结果] 连接成功=" << connOk << "/" << nconn << "\n";
    std::cout << "[结果] 请求成功=" << okReq << "/" << nconn * repeat
              << "  请求失败=" << failReq << "\n";
    std::cout << "[结果] 总耗时=" << dur << "s  连接速率≈"
              << (dur > 0 ? (int)(nconn / dur) : 0) << "/s  请求速率≈"
              << (dur > 0 ? (int)(okReq / dur) : 0) << "/s\n";
    return 0;
}
